pkg/agent/monitortool/monitor.go

```diff
@@ -0,0 +1,449 @@
+func (m *NodeLatencyMonitor) pingAll(ipv4Socket, ipv6Socket net.PacketConn) {
+	klog.V(4).InfoS("Pinging all Nodes")
+	nodeIPs := m.latencyStore.ListNodeIPs()
+	for _, toIP := range nodeIPs {
+		if toIP.To4() != nil && ipv4Socket != nil {
+			if err := m.sendPing(ipv4Socket, toIP); err != nil {
+				klog.ErrorS(nil, "Cannot send ICMP message to Node IP because socket is not initialized for IPv4", "IP", toIP)
+			}
+		} else if toIP.To16() != nil && ipv6Socket != nil {
+			if err := m.sendPing(ipv6Socket, toIP); err != nil {
+				klog.ErrorS(nil, "Cannot send ICMP message to Node IP because socket is not initialized for IPv6", "IP", toIP)
+			}
+		} else {
+			klog.ErrorS(nil, "Cannot send ICMP message to Node IP because socket is not initialized for IP family", "IP", toIP)
+		}
+	}
+	klog.V(4).InfoS("Done pinging all Nodes")
+}
+
+// Run starts the NodeLatencyMonitor.
+func (m *NodeLatencyMonitor) Run(stopCh <-chan struct{}) {
+	go m.nodeLatencyMonitorInformer.Informer().Run(stopCh)
+	go m.nodeInformer.Informer().Run(stopCh)
+	go m.monitorLoop(stopCh)
+
+	<-stopCh
+}
+
+// monitorLoop is the main loop to monitor the latency of the Node.
+func (m *NodeLatencyMonitor) monitorLoop(stopCh <-chan struct{}) {
+	klog.V(4).InfoS("NodeLatencyMonitor is running")
+	// Low level goroutine to handle ping loop
+	var ticker *time.Ticker
+	var tickerCh <-chan time.Time
+	var ipv4Socket, ipv6Socket net.PacketConn
+	var err error
+
+	defer func() {
+		if ipv4Socket != nil {
+			ipv4Socket.Close()
+		}
+		if ipv6Socket != nil {
+			ipv6Socket.Close()
+		}
+		if ticker != nil {
+			ticker.Stop()
+		}
+	}()
+
+	// Update current ticker based on the latencyConfig
+	updateTicker := func(interval time.Duration) {
+		if ticker != nil {
+			ticker.Stop() // Stop the current ticker
+		}
+		ticker = time.NewTicker(interval)
+		tickerCh = ticker.C
+	}
+
+	klog.V(4).InfoS("NodeLatencyMonitor is running2")
+	wg := sync.WaitGroup{}
+	// Start the pingAll goroutine
+	for {
+		select {
+		case <-tickerCh:
+			// Try to send pingAll signal
+			m.pingAll(ipv4Socket, ipv6Socket)
+		case <-stopCh:
+			return
+		case <-m.latencyConfigChanged:
+			// Start or stop the pingAll goroutine based on the latencyConfig
+			if m.latencyConfig.Enable {
+				// If the Nodes is empty(after deleting the CRD), we need to wait for the NodeInformer to populate the Nodes.
+				if len(m.latencyStore.ListNodeIPs()) == 0 {
+					nodes, err := m.nodeInformer.Lister().List(labels.Everything())
+					if err != nil {
+						klog.ErrorS(err, "Failed to list Nodes")
+					}
+					m.latencyStore.addNodes(nodes)
+				}
+
+				// latencyConfig changed
+				updateTicker(m.latencyConfig.Interval)
+
+				// If the recvPing socket is closed, restart it
+				// In case of IPv4-only or IPv6-only, we need to check the socket status,
+				// and restart it if it is closed(CRD is deleted).
+				if ipv4Socket == nil && m.isIPv4Enabled {
+					// Create a new socket for IPv4 when the gatewayConfig is IPv4-only
```

Comment:
wrong: this is not specific to IPv4-only. It applies to dual-stack as well.

IPv4-only means that the cluster only uses IPv4 networking. Dual-stack means that the cluster supports both IPv4 and IPv6 networking.
This logic applies whenever the cluster supports IPv4 (IPv4 is enabled), which can be the IPv4-only case or the dual-stack case.
Additionally, why is there a reference to `gatewayConfig` here?

Reply:
Ok, I will modify the comments here.
Regarding the `gatewayconfig`, it was previously set against the net stack checksum, I will modify it as well.